lib/chef/provider/package/apt.rb

```diff
@@ -255,6 +273,81 @@ def package_data_for(pkg)
           }
         end
 
+        # Helper to construct Hash of names-to-package-information.
+        #
+        # @return [Hash] Mapping of package names to package information
+        def name_pkginfo
+          @name_pkginfo ||=
+            begin
+              pkginfos = resolved_source_array.map do |src|
+                logger.trace("#{new_resource} checking #{src} dpkg status")
+                status = shell_out("dpkg-deb", "-W", src)
+                status.stdout
+              end
+              Hash[*package_name_array.zip(pkginfos).flatten]
+            end
+        end
+
+        def name_package_name
+          @name_package_name ||= name_pkginfo.transform_values { |v| v ? v.split("\t")[0] : nil }
+        end
+
+        # Return package names from the candidate source file(s).
+        #
+        # @return [Array] Array of actual package names read from the source files
+        def get_package_name
+          package_name_array.map { |name| name_package_name[name] }
+        end
+
+        def read_current_version_of_package(package_name)
+          logger.trace("#{new_resource} checking install state of #{package_name}")
+          status = shell_out!("dpkg", "-s", package_name, returns: [0, 1])
+          package_installed = false
+          status.stdout.each_line do |line|
+            case line
+            when /^Status: deinstall ok config-files/.freeze
```

Comment:
We especially don't want to duplicate these regexes. The pattern we use for defining these over in https://github.com/chef/chef/blob/main/lib/chef/provider/package/dpkg.rb#L28-L30 is intentional. But duplicating these, if there was a bug we needed to fix in one provider we'd have to remember that we duplicated this code in another provider and fix the same regex. We'd likely forget that and miss the fix.